NewareNDA/NewareNDAx.py

```diff
@@ -1,609 +1,588 @@
+
+    # Convert to pandas, change timestamp to local timezone
+    df = df.to_pandas()
+    tz = datetime.now().astimezone().tzinfo
+    df["Timestamp"] = pd.to_datetime(df["Timestamp"]).dt.tz_localize("UTC").dt.tz_convert(tz)
+    return df
+
+def extract_and_read_ndc(zf: zipfile.ZipFile, filename: str, tmpdir: str) -> tuple[str, pd.DataFrame | None]:
+    """Extract .ndc from a zipfile and reads it into a DataFrame."""
+    if filename in zf.namelist():
+        file_path = zf.extract(filename, path=tmpdir)
+        return filename, read_ndc(file_path)
+    return filename, None
+
+def _data_interpolation(df):
+    """
+    Some ndax from from BTS Server 8 do not seem to contain a complete dataset.
+    This helper function fills in missing times, capacities, and energies.
+    """
+    # Get time by forward filling differences
+    df = df.with_columns([
+        pl.col("Time").is_not_null().alias("nan_mask"),
+        pl.col("Time").is_not_null().cum_sum().shift(1).fill_null(0).alias("group_idx"),
+        pl.col("dt", "Time", "uts", "Charge_Capacity(mAh)", "Discharge_Capacity(mAh)",
+           "Charge_Energy(mWh)", "Discharge_Energy(mWh)").fill_null(strategy="forward"),
+    ])
+
+    df = df.with_columns([
+        (pl.col("dt").cum_sum().over("group_idx") * (~pl.col("nan_mask"))).alias("cdt"),
+        (pl.col("dt") * pl.col("Current(mA)") / 3600).cum_sum().over("group_idx").alias("capacity"),
+    ])
+
+    df = df.with_columns([
+        (pl.col("Time") + pl.col("cdt")).alias("Time"),
+        (pl.col("uts") + pl.col("cdt")).alias("uts"),
+        (pl.col("Charge_Capacity(mAh)") + pl.col("capacity").clip(lower_bound=0)).alias("Charge_Capacity(mAh)"),
+        (pl.col("Discharge_Capacity(mAh)") + pl.col("capacity").clip(upper_bound=0)).alias("Discharge_Capacity(mAh)"),
+        (pl.col("Charge_Energy(mWh)") + pl.col("capacity").clip(lower_bound=0) * pl.col("Voltage")).alias("Charge_Energy(mWh)"),
+        (pl.col("Discharge_Energy(mWh)") + pl.col("capacity").clip(upper_bound=0) * pl.col("Voltage")).alias("Discharge_Energy(mWh)"),
+    ])
+
+    # Sanity checks
+    if (df["uts"].diff() < 0).any():
+        logger.warning(
+            "IMPORTANT: This ndax has negative jumps in the 'Timestamp' column! "
+            "This can sometimes happen in the ndax file itself. "
+            "Use the 'Time' column for analysis.",
+        )
+
+    return df
+
+
+def read_ndc(file):
+    """
+    Function to read electrochemical data from a Neware ndc binary file.
+
+    Args:
+        file (str): Name of an .ndc file to read
+    Returns:
+        df (pd.DataFrame): DataFrame containing all records in the file
+        aux_df (pd.DataFrame): DataFrame containing any temperature data
+    """
+    with open(file, "rb") as f, mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ) as mm:
+        # Get ndc file version and filetype
+        [ndc_filetype] = struct.unpack("<B", mm[0:1])
+        [ndc_version] = struct.unpack("<B", mm[2:3])
```

Comment:
Would you mind re-adding the debug log here? I think that one has been useful to quickly see what version a file has

```
logger.debug(f"NDC version: {ndc_version} filetype: {ndc_filetype}")
```